cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+        GetErrorStr());
+
+    return -1;
+}
+
+static Rlist *ReadDataFromPackageScript(const IOData *io)
+{
+    char buff[CF_BUFSIZE] = {0};
+    
+    Buffer *data = BufferNew();
+    if (!data)
+    {
+        Log(LOG_LEVEL_VERBOSE, 
+            "Unable to allocate buffer for handling package module responses.");
+        return NULL;
+    }
+    
+    int timeout_seconds_left = PACKAGE_PROMISE_SCRIPT_TIMEOUT_SEC;
+    
+    while(!IsPendingTermination() && timeout_seconds_left > 0)
+    {
+        int fd = IsReadWriteReady(io, PACKAGE_PROMISE_TERMINATION_CHECK_SEC);
+        
+        if (fd < 0)
+        {
+            Log(LOG_LEVEL_VERBOSE, 
+                "Error reading data from package module: %s",
+                GetErrorStr());
+            return NULL;
+        }
+        else if (fd == io->read_fd)
+        {
+            ssize_t res = read(fd, buff, sizeof(buff) - 1);
+            if (res == -1)
+            {
+                if (errno == EINTR)
+                {
+                    continue;
+                }
+                else
+                {
+                    Log(LOG_LEVEL_ERR,
+                        "Unable to read output from package module: %s",
+                        GetErrorStr());
+                    BufferDestroy(data);
+                    return NULL;
+                }
+            }
+            else if (res == 0) /* reached EOF */
+            {
+                break;
+            }
+            Log(LOG_LEVEL_DEBUG, "Data read from package module: %zu [%s]",
+                res, buff);
+
+            BufferAppendString(data, buff);
+            memset(buff, 0, sizeof(buff));
+        }
+        else if (fd == 0) /* timeout */
+        {
+            timeout_seconds_left -= PACKAGE_PROMISE_TERMINATION_CHECK_SEC;
+            continue;
+        }
+    }
+    
+    char *read_string = BufferClose(data);
+    Rlist *response_lines = RlistFromSplitString(read_string, '\n');
+    free(read_string);
+    
+    return response_lines;
+}
+
+static int WriteScriptData(const char *data, IOData *io)
+{
+    /* If there is nothing to write close writing end of pipe. */
+    if (data == NULL || strlen(data) == 0)
+    {
+        if (io->write_fd >= 0)
+        {
+            close(io->write_fd);
+            io->write_fd = -1;
+        }
+        return 0;
+    }
+    
+    ssize_t wrt = write(io->write_fd, data, strlen(data));
+    
+    /* Make sure to close write_fd after sending all data. */
+    if (io->write_fd >= 0)
+    {
+        close(io->write_fd);
+        io->write_fd = -1;
+    }
+    return wrt;
+}
+
+int WriteDataToPackageModule(const char *args, const char *data,
+                             const PackageModuleWrapper *wrapper)
+{
+    char *command = StringFormat("%s %s", wrapper->path, args);
+    IOData io = cf_popen_full_duplex(command, false);
+    free(command);
+    
+    if (io.write_fd == 0 || io.read_fd == 0)
```

Comment:
0 is a bad sentinel for 'bad descriptor' as it's actually a valid descriptor; if stdin has ever been close()d, it *shall* get used.